### PR TITLE
chore(chain-adapters): cleanup coinselect use

### PR DESCRIPTION
**Pull request by elmutt**
- cleanup how we use coinselect with buildTransaction and getFees
- add support for oprReturnData

### CODE REVIEW

**Review by kaladinlight — APPROVED** [review]

File: packages/chain-adapters/src/bitcoin/BitcoinChainAdapter.ts [hunk 1]
```diff
@@ -176,10 +177,22 @@ export class ChainAdapter
...
+      // op_return is ultimately added to the transaction upon signing but not here
+      // Imperfect because the assumption of 300 bytes for a standard tx could vary with different inputs/outputs
+      // Another solution would be to write our own coinselect algorithm
+      const standardTxBytes = 300
+      const opReturnTxBytes = standardTxBytes + (opReturnData?.length ?? 0)
+      const adjustedSatsPerByte = bn(satoshiPerByte)
+        .times(bn(opReturnTxBytes).div(standardTxBytes))
+        .dp(0)
+        .toString()
+
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/bitcoin/BitcoinChainAdapter.ts:195]: it seems like we should do all of this within bn, and we're making a string to immediately convert to a number into coinselect?

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]

**Comment by elmutt** [comment]
put back as a draft.

Looking a better approach after conversation with @kaladinlight

File: packages/chain-adapters/src/bitcoin/BitcoinChainAdapter.test.ts [hunk 2]
```diff
@@ -265,7 +265,6 @@ describe('BitcoinChainAdapter', () => {
         value: '400',
         wallet,
         chainSpecific: {
-          opReturnData: 'nm, u',
```
Comment (elmutt) [inline packages/chain-adapters/src/bitcoin/BitcoinChainAdapter.test.ts:268]: Tests concerning opReturnData will come in a subsequent PR when the rest of opReturnData is implemented

File: packages/chain-adapters/src/bitcoin/BitcoinChainAdapter.ts [hunk 3]
```diff
@@ -316,43 +312,28 @@ export class ChainAdapter
...
+      sendMax
     }
 
+    const { fee: fastFee } = utxoSelect({
+      ...utxoSelectInput,
+      satoshiPerByte: fastPerByte
+    })
+    const { fee: averageFee } = utxoSelect({
+      ...utxoSelectInput,
+      satoshiPerByte: fastPerByte
```
Comment (kaladinlight) [inline packages/chain-adapters/src/bitcoin/BitcoinChainAdapter.ts:330]:
```suggestion
      satoshiPerByte: averagePerByte
```

File: packages/chain-adapters/src/bitcoin/BitcoinChainAdapter.ts [hunk 4]
```diff
@@ -316,43 +312,28 @@ export class ChainAdapter
...
-      )
-      fastFee = fast
-      averageFee = average
-      slowFee = slow
+    const utxoSelectInput = {
+      to,
+      value,
+      opReturnData,
+      utxos,
+      satoshiPerByte: fastPerByte,
```
Comment (kaladinlight) [inline packages/chain-adapters/src/bitcoin/BitcoinChainAdapter.ts:320]:
```suggestion
```

File: packages/chain-adapters/src/bitcoin/utxoSelect.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    value,
+    to,
+    satoshiPerByte,
+    opReturnData,
+    sendMax
+  })
+  throw new Error('not implemented')
+}
+
+export const utxoSelect = (input: {
```
Comment (cjthompson) [inline packages/chain-adapters/src/bitcoin/utxoSelect.ts:62]: Can you add a doc-block with an explanation about what this is doing? I assume it has something to do with picking which UTXO to include in a transaction?

File: packages/chain-adapters/src/bitcoin/utxoSelect.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    return coinSelect<MappedUtxos, bitcoin.Recipient>(
+      mappedUtxos,
+      [{ value: Number(value), address: to }],
+      Number(satoshiPerByte)
+    )
+  }
+}
+
+// TODO this will be implemented in another PR to keep concerns separated and make review easier
+const handleOpReturnTx = ({
```
Comment (kaladinlight) [inline packages/chain-adapters/src/bitcoin/utxoSelect.ts:36]: May be easy enough to just push on the opReturn output if not undefined and just have a single function unless there are more differences that arise

File: packages/chain-adapters/src/bitcoin/utxoSelect.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
+/* eslint-disable no-console */
+import * as unchained from '@shapeshiftoss/unchained-client'
+import coinSelect from 'coinselect'
+import split from 'coinselect/split'
+
+import { bitcoin } from '../types'
+
+export type MappedUtxos = Omit<unchained.bitcoin.Utxo, 'value'> & { value: number }
+
+const handleStandardTx = ({
```
Comment (cjthompson) [inline packages/chain-adapters/src/bitcoin/utxoSelect.ts:9]: the name `handle` is usually used for functions that handle events. Is there a more descriptive name of what this is actually doing?  I'm not entire sure what it's doing.

File: packages/chain-adapters/src/bitcoin/utxoSelect.ts (hunk 6 above)
Comment (elmutt) [reply packages/chain-adapters/src/bitcoin/utxoSelect.ts:36]: ya true that may be the case but for now this keeps it separate and its east enough to merge back into a single function if we find thats better

**Comment by elmutt** [comment]
> Should these changes go into the `UTXOBaseAdapter`?

No. getFeeData and buildSendTransaction are abstract and not implemented in the base utxo adapter.

**Review by kaladinlight — APPROVED** [review]

File: packages/chain-adapters/src/bitcoin/utxoSelect.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaladinlight) [reply packages/chain-adapters/src/bitcoin/utxoSelect.ts:62]: what are we testing? the library func that already has tests?

**Review by DaoDev44 — APPROVED** [review]

**Review by cjthompson — DISMISSED** [review]

File: packages/chain-adapters/src/bitcoin/utxoSelect/utxoSelect.test.ts [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        {
+          value: 400,
+          address: 'bc1qppzsgs9pt63cx9x994wf4e3qrpta0nm6htk9v4'
+        },
+        { value: 31335 }
+      ],
+      fee: 226
+    }
+    const result = utxoSelect({ ...utxoSelectInputStandard, sendMax: false })
+    console.log('result is', result)
```
Comment (kaladinlight) [inline packages/chain-adapters/src/bitcoin/utxoSelect/utxoSelect.test.ts:58]:
```suggestion
```

File: packages/chain-adapters/src/bitcoin/utxoSelect/utxoSelect.test.ts [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    const result = utxoSelect({ ...utxoSelectInputStandard, sendMax: true })
+
+    expect(result).toEqual(expectedOutput)
+  })
+
+  it('should return correct inputs and outputs and fee for a sendmax tx with opReturnData', () => {
+    const result = utxoSelect({ ...utxoSelectInputOpReturn, sendMax: true })
+
+    console.log('result is', JSON.stringify(result, null, 2))
+
```
Comment (kaladinlight) [inline packages/chain-adapters/src/bitcoin/utxoSelect/utxoSelect.test.ts:100]:
```suggestion
```

**Review by kaladinlight — APPROVED** [review]

**Review by kaladinlight — DISMISSED** [review]

**Review by kaladinlight — APPROVED** [review]

**Review by kaladinlight — APPROVED** [review]

**Review by kaladinlight — APPROVED** [review]

**Pull request merged**